### PR TITLE
Bug/441 0 validation

### LINKED ISSUE #441: O validation items
**Describe the bug**
```
n. | /home/peter/Github/nowcasting_dataset/nowcasting_dataset/manager.py#L185
DEBUG:nowcasting_dataset.manager:Creating 10 batches x 32 examples per batch = 320 examples for validation.
Traceback (most recent call last):
  File "/home/peter/Github/nowcasting_dataset/scripts/prepare_ml_data.py", line 79, in <module>
    main()
  File "/home/peter/miniconda3/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/peter/miniconda3/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/peter/miniconda3/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/peter/miniconda3/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/peter/Github/nowcasting_dataset/nowcasting_dataset/utils.py", line 155, in inner_func
    return func(*args, **kwargs)
  File "/home/peter/Github/nowcasting_dataset/scripts/prepare_ml_data.py", line 71, in main
    manager.create_files_specifying_spatial_and_temporal_locations_of_each_example_if_necessary()
  File "/home/peter/Github/nowcasting_dataset/nowcasting_dataset/manager.py", line 189, in create_files_specifying_spatial_and_temporal_locations_of_each_example_if_necessary
    df_of_locations = self.sample_spatial_and_temporal_locations_for_examples(
  File "/home/peter/Github/nowcasting_dataset/nowcasting_dataset/manager.py", line 290, in sample_spatial_and_temporal_locations_for_examples
    shuffled_t0_datetimes = np.random.choice(t0_datetimes, size=n_examples)
  File "mtrand.pyx", line 915, in numpy.random.mtrand.RandomState.choice
ValueError: 'a' cannot be empty unless no samples are taken
```

**To Reproduce**
run script preapare_ml_data.py

### PR DESCRIPTION
**Pull request by peterdudfield**
# Pull Request

## Description

fix bug when there are zero batches for validation

Fixes #441 

## How Has This Been Tested?

- normal unittests
- addiontal unittests

- [ ] No
- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: nowcasting_dataset/manager.py [hunk 1]
```diff
@@ -287,6 +287,32 @@ def sample_spatial_and_temporal_locations_for_examples(
             Each row of each the DataFrame specifies the position of each example, using
             columns: 't0_datetime_UTC', 'x_center_OSGB', 'y_center_OSGB'.
         """
+
+        if n_examples == 0:
+            logger.debug(
+                f"Requested {n_examples} so will" f"return zero spatial and temporal locations"
```
Comment (peterdudfield) [inline nowcasting_dataset/manager.py:293]: fix

File: nowcasting_dataset/manager.py [hunk 2]
```diff
@@ -287,6 +287,32 @@ def sample_spatial_and_temporal_locations_for_examples(
...
+            )
+            return pd.DataFrame(
+                {
+                    "t0_datetime_UTC": [],
+                    "x_center_OSGB": [],
+                    "y_center_OSGB": [],
+                }
+            )
+
+        if (len(t0_datetimes) == 0) and (n_examples > 0):
```
Comment (peterdudfield) [inline nowcasting_dataset/manager.py:303]: will refactor to put these together

**Comment by JackKelly** [comment]
So... I think that perhaps the root cause of this issue is that we allow users to request a non-zero number of validation batches, even though the splitting function returns zero `t0_datetimes` for validation...

**Comment by JackKelly** [comment]
So I wonder if the `model.py` code should enforce that the user cannot request validation batches when the splitting function will refuse to return validation batches?

**Comment by peterdudfield** [comment]
> So I wonder if `model.py` should enforce that the user cannot request validation batches when the splitting function will return zero validation batches?

yea maybe this could be part of
https://github.com/openclimatefix/nowcasting_dataset/issues/426

I hope this PR make it not fail, but produce a warning anyway, for this kinda of issue

**Comment by JackKelly** [comment]
Ah, yeah, good point... I think I'm close to having a fix which sets n_validation_batches to zero...

**Comment by peterdudfield** [comment]
Ill close this, as I think this has been solved, and its out of date

**Pull request closed without merging**